lib/docker_manager/upgrader.rb

```diff
@@ -84,13 +84,22 @@ def upgrade
         run "cd #{repo.path} && git branch -u origin/main main"
         run("cd #{repo.path} && git reset --hard HEAD@{upstream}")
       else
-        run("cd #{repo.path} && git fetch --tags --force && git reset --hard HEAD@{upstream}")
+        run("cd #{repo.path} && git fetch --tags --prune-tags --prune --force")
+
+        if repo.detached_head?
+          run("cd #{repo.path} && git reset --hard")
+          run("cd #{repo.path} && git -c advice.detachedHead=false checkout #{repo.tracking_ref}")
+        else
+          run("cd #{repo.path} && git reset --hard HEAD@{upstream}")
+        end
       end
 
       percent(20 * (index + 1) / @repos.size)
     end
 
-    run("bundle install --deployment --jobs 4 --without test development")
+    run("bundle config --local deployment true")
+    run("bundle config --local without 'development test'")
+    run("bundle install --retry 3 --jobs 4")
     run("yarn install --production")
```

Comment:
cross-linking https://github.com/discourse/discourse_docker/pull/708

cc: @davidtaylorhq 

```diff
-    run("yarn install --production")
+    run("yarn install")
```

Reply:
:+1: yep, we'll need to make this change eventually, but let's do it as a separate PR

Reply:
Hmm, is it safe to replace this? What about the other commands and options from the PR you linked?

```
yarn install --frozen-lockfile && yarn cache clean
```

I think I'll leave it for now. Maybe @davidtaylorhq can remove it in the future when https://github.com/discourse/discourse_docker/pull/708 gets merged?